### ISSUE TITLE
Scrolling a scrollable widget with Switch Control on iOS is not very comfortable

### ISSUE THREAD

**Issue by chunhtai** [flutter/flutter#37974]
design doc: [flutter.dev/go/ios-switch-controll-scrolling](http://flutter.dev/go/ios-switch-controll-scrolling)

## Steps to Reproduce

```
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'List playground',
      home: TestPage(),
    );
  }
}

class TestPage extends StatefulWidget {
  @override
  State<StatefulWidget> createState() => _TestPageState();
}

class _TestPageState extends State<TestPage> {
  List<String> items = ['1', '2', '3', '4', '5'];
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: SizedBox(
          width: 300.4,
          height: 50.0,
          child: Directionality(
            textDirection: TextDirection.ltr,
            child: CustomScrollView(
              slivers: <Widget>[
                SliverFixedExtentList(
                  itemExtent: 22.2,
                  delegate: SliverChildBuilderDelegate(
                      (BuildContext context, int index) {
                      return TextWidget(
                        items[index],
                      );
                    },
                    childCount : items.length,
                  ),
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}
class TextWidget extends StatefulWidget {
  const TextWidget(this.data);
  final String data;


  @override
  TextWidgetState createState() => TextWidgetState();
}

class TextWidgetState extends State<TextWidget>{

  @override
  void dispose() {
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Text(widget.data);
  }
}
```

1. turn on switch control
2. try to select the text item

It should show the scroll options, but it only has tap. 

## Logs

```
```

```
```

**Comment by mehmetf** [comment]
I just learned from a Google team that their app was nearly blocked from release on iOS due to this issue. They got a temporary exception.

**Comment by Hixie** [comment]
@goderbauer can you elucidate on what you were telling me earlier about what we need to be able to fix this?

**Comment by chunhtai** [comment]
Last time i check there is no available API in IOS to make a UI view to be consider scrollable in accessibility. The only way to make it accessibly scrollable in IOS is to use native UIScrollView.

**Comment by goderbauer** [comment]
@mehmetf Just be sure that we are talking about the same thing: Can you confirm that the blocker was that Flutter doesn't show this "Scroll down/up" option in the Switch Access action menu? Is there an internal bug that has more information about what required the temporary exception? 

![IMG_0001](https://user-images.githubusercontent.com/1227763/79388418-2683aa00-7f22-11ea-8ce8-02fd2198bfb4.PNG)

@Hixie If that is indeed the issue here, the problem is that iOS doesn't seem to offer an official API to make that option appear in the menu as pointed out by @chunhtai above. @chunhtai Didn't you also ask on StackOverflow if there's a way to implement this? Can you maybe link to that question?

One thing we could try is reverse engineer how UIScrollView interacts with the iOS a11y subsystem to figure out if we can maybe steal one of its a11y traits and use it for our scroll view. We'd have to implement that in the iOS a11y bridge (/cc @xster).

**Comment by chunhtai** [comment]
I have tried to ask this question in stackoverflow https://stackoverflow.com/questions/57466350/ios-switch-control-scrolling-in-opengl-backed-uiview
 but got no response

**Comment by gaaclarke** [comment]
I tried a few things to reverse engineer this:
1) I zero'd out the `accessibiilityTraits` property of a UISwitch and the UIScrollView, this causes VoiceOver to be wrong, but the action menu still has (tap, scroll).
1) I tried to use NSObjCMessageLoggingEnabled to log the messages being sent to the UISwitch when the action menu is displayed.  There are problems [getting the logs on device](https://stackoverflow.com/questions/25515730/log-objective-c-message-sends-on-a-device)  In order to get it working on device you have to fiddle with non-exported symbols.  I couldn't get it working and accessibility doesn't work on simulator.
1) I tried to hack objc_msgSend to log all methods sent to the UISwitch.  I tried using NSProxy and overriding `forwardInvocation`. UIKit seems to replace my proxy object or crash, similar to the problem [this person had](https://stackoverflow.com/questions/5061223/anybody-successful-with-nsproxy-of-uiview-e-g-uilabel)

**Comment by mehmetf** [comment]
The exception currently exists as an internal document. In fact there are a couple of complications that we need to resolve first before filing an internal bug. I will drive this with the client. Removing the critical tag for now.

**Comment by gaaclarke** [comment]
In an attempt to figure out how that menu is propagated, I tried placing a breakpoint on `objc_msgSend` that doesn't pause execution with a conditional `$x0 == <pointer to uiswitch>` printing out `(char*)$x1`.  Even with the conditional, the breakpoint bogs down the system to the point where the UI isn't responsive and the Switch Control never finds the UISwitch.

**Comment by goderbauer** [comment]
@gaaclarke Just to be sure there's no misunderstanding: I don't think this is going to be anything on the UISwitch widget. The "scroll down" button in the screenshot above is somehow produced by the UIScrollView surrounding the element that currently has accessibility focus (it's just coincidence that focus is on a UISwitch in my example above). 

I don't know how the UIScrollView produces the button, though. Maybe iOS internally walks up the hierarchy from the currently focused element to see if there are any UIScrollViews (or hopefully for us anything with an UIScrollView a11y trait) and if there are it adds that button? Idk...

**Comment by gaaclarke** [comment]
@goderbauer I was thinking that too.  The techniques I tried would have worked for that as well but I couldn't get any of them working.  I didn't mention it but I tried zeroing out the accessibilityTraits of the UIScrollView but that button still showed up in the action menu.

**Comment by goderbauer** [comment]
@gaaclarke Thanks for looking into it.

Just for the record: There is a way to still scroll in Flutter with Switch Control on iOS: 
* Select an item inside a scrollable
* Switch to the second page of the Switch Control menu that appeared
* Select "Gestures" > "Flick"
* Select the direction in which you want to scroll

**Comment by goderbauer** [comment]
@j-sid had an interesting idea for a workaround: We could use `UIAccessibilityCustomAction`s to add custom items to the Switch Control menu of every element within a scrollable that can be used to scroll said scrollable.

Basically, for each SemanticsObject in the iOS bridge we would have check whether it is in a scrollable. If it is, we figure out what direction the scrollable can scroll in and then add the appropriate `UIAccessibilityCustomAction`s for those scroll directions to the SemanticsObject. We should only do this if Switch Access is enabled (Those custom actions may otherwise confuse VoiceOver users, which don't need them since Flutter has proper scroll support there). 

Open Questions:
* How do we internationalize the label for the custom action on the engine side? Do we somehow need to pipe the internationalized labels in from the framework side?
* What does the icon for these custom actions look like in the Switch Access menu? As far as I can see, there is no way in the API to provide a custom icon to match the icon that iOS produces natively. (This probably wouldn't be a deal breaker).

**Comment by goderbauer** [comment]
Looks like starting in iOS 14 we will even be able to provide a custom icon for those actions: https://developer.apple.com/videos/play/wwdc2020/10019/

**Comment by chunhtai** [comment]
bumping to p3 because this is what we want to look into in this quarter

**Comment by TahaTesser** [comment]
<details> 
<summary>Code Sample</summary> 

```
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'List playground',
      home: TestPage(),
    );
  }
}

class TestPage extends StatefulWidget {
  @override
  State<StatefulWidget> createState() => _TestPageState();
}

class _TestPageState extends State<TestPage> {
  List<String> items = ['1', '2', '3', '4', '5'];
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: SizedBox(
          width: 300.4,
          height: 50.0,
          child: Directionality(
            textDirection: TextDirection.ltr,
            child: CustomScrollView(
              slivers: <Widget>[
                SliverFixedExtentList(
                  itemExtent: 22.2,
                  delegate: SliverChildBuilderDelegate(
                    (BuildContext context, int index) {
                      return TextWidget(
                        items[index],
                      );
                    },
                    childCount: items.length,
                  ),
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}

class TextWidget extends StatefulWidget {
  const TextWidget(this.data);
  final String data;

  @override
  TextWidgetState createState() => TextWidgetState();
}

class TextWidgetState extends State<TextWidget> {
  @override
  void dispose() {
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Text(widget.data);
  }
}

``` 

</details> 

<details> 
<summary>flutter doctor -v</summary> 

```
[✓] Flutter (Channel dev, 1.21.0-1.0.pre, on Mac OS X 10.15.6 19G73, locale
    en-GB)
    • Flutter version 1.21.0-1.0.pre at /Users/taha/Code/flutter_dev
    • Framework revision f25bd9c55c (8 days ago), 2020-07-14 20:26:01 -0400
    • Engine revision 99c2b3a245
    • Dart version 2.9.0 (build 2.9.0-21.0.dev 20bf2fcf56)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at /Users/taha/Code/sdk
    • Platform android-29, build-tools 29.0.3
    • ANDROID_HOME = /Users/taha/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.6)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.6, Build version 11E708
    • CocoaPods version 1.9.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 47.1.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.47.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.12.2

[✓] Connected device (4 available)
    • SM M305F (mobile) • 32003c30dc19668f • android-arm64  • Android 10 (API
      29)
    • macOS (desktop)   • macos            • darwin-x64     • Mac OS X 10.15.6
      19G73
    • Web Server (web)  • web-server       • web-javascript • Flutter Tools
    • Chrome (web)      • chrome           • web-javascript • Google Chrome
      84.0.4147.89

• No issues found!
```

</details>

**Comment by chunhtai** [comment]
note to myself:

1. The scroll direction is inverse of flutter. The `Scroll Down` in switch controller means scroll up in flutter (unnature direction)

2. When there are nested scroll views one within another, the inner one always scrolls first.

**Comment by chunhtai** [comment]
We should also fix this https://github.com/flutter/flutter/issues/64279, otherwise the switch control scroll will be pretty confusing
